crates/telio-wg/src/no_link_detection.rs

```diff
@@ -0,0 +1,522 @@
+            None
+        } else if !matches!(node_state, NodeState::Connected) {
+            Some(LinkState::Down)
+        } else {
+            Some(link_state)
+        };
+
+        NoLinkDetectionUpdateResult {
+            should_notify,
+            link_state,
+        }
+    }
+}
+
+pub struct NoLinkDetectionUpdateResult {
+    pub should_notify: bool,
+    pub link_state: Option<LinkState>,
+}
+
+/// No link detection mechanism config
+#[derive(Clone)]
+enum Config {
+    /// No link detection mechanism is disabled
+    Disabled,
+    /// No link detection mechanism is enabled
+    /// Which means that events with link_state will be emitted
+    Enabled {
+        /// Configurable rtt
+        rtt: Duration,
+    },
+}
+
+impl From<Option<FeatureNoLinkDetection>> for Config {
+    fn from(value: Option<FeatureNoLinkDetection>) -> Self {
+        match value {
+            None => Config::Disabled,
+            Some(FeatureNoLinkDetection { rtt_seconds: None }) => Config::Enabled {
+                rtt: DEFAULT_CONFIGURABLE_RTT,
+            },
+            Some(FeatureNoLinkDetection {
+                rtt_seconds: Some(rtt),
+            }) => Config::Enabled {
+                rtt: Duration::from_secs(rtt),
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+struct NodeInfo {
+    last_tx_timestamp: Instant,
+    last_rx_timestamp: Instant,
+    last_tx_bytes: u64,
+    last_rx_bytes: u64,
+    last_link_state: LinkState,
+}
+
+impl Default for NodeInfo {
+    fn default() -> Self {
+        let now = Instant::now();
+        NodeInfo {
+            last_tx_timestamp: now,
+            last_rx_timestamp: now,
+            last_tx_bytes: 0,
+            last_rx_bytes: 0,
+            last_link_state: LinkState::Down,
+        }
+    }
+}
+
+impl NodeInfo {
+    fn compute_new_node_info(
+        &self,
+        new_rx: Option<u64>,
+        new_tx: Option<u64>,
+        rtt: Duration,
+    ) -> Self {
+        let new_rx = new_rx.unwrap_or(self.last_rx_bytes);
+        let new_tx = new_tx.unwrap_or(self.last_tx_bytes);
+
+        let now = Instant::now();
+
+        let (rx_bytes, rx_ts) = if new_rx > self.last_rx_bytes {
+            (new_rx, now)
+        } else {
+            (self.last_rx_bytes, self.last_rx_timestamp)
+        };
+
+        let (tx_bytes, tx_ts) = if new_tx > self.last_tx_bytes {
+            // Ignore the keepalive messages
+            // This only works in a limited fashion because we poll every second and
+            // the detection might not work as intended if there are more packets within that time window.
+            // On the sender side persistent-keepalives might interfere as those are emitted regardless the activity,
+            // meaning for example if persistent-keepalive is set at 5seconds then it might interfere with the
+            // passive-keepalive which is a constant of 10seconds.
+            let tx_ts = if new_tx - self.last_tx_bytes == 32 {
```

Comment:
Can you elaborate on what `32` means here?

Reply:
32 is the size of a keepalive packet. I'll make it a constant.

Reply:
We want to ignore them from the sending perspective

Reply:
Ok. But is it possible that some other packet has 32 bytes? Thus breaking our logic here?